daemon/src/common/Profiler.cpp

```diff
@@ -0,0 +1,201 @@
+DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+===========================================================================
+*/
+
+#include "Profiler.h"
+
+#include "Profiler.h"
+#include "../engine/framework/LogSystem.h"
+
+#include "../engine/framework/CommandSystem.h"
+#include "../engine/framework/BaseCommands.h"
+#include "FileSystem.h"
+#include <fstream>
+
+
+namespace Profiler{
+
+    /*
+         * Function execution times are saved in the samples vector, every element
+         * contains the name of the function, the type of sample (function start,
+         * function end, and if there's a new frame), and a timestamp.
+         */
+    std::vector <Profiler::Point>                   samples;
+    thread_local std::vector <Profiler::Point>      localSamples;
+    std::mutex mtx;
+
+    Sys::SteadyClock::time_point                    startTime;
+    bool                                            enabled    = false; ///TODO those bools are silly
+    bool                                            nextFrame  = false;
+    bool                                            shouldStop = false;
+
+
+
+    /// Called at the end of every frame
+    void Frame(){
+
+        if(nextFrame){
+            enabled=true;
+            shouldStop=true;
+            nextFrame=false;
+            startTime   = Sys::SteadyClock::now();
+            return;
+        }
+
+        if(shouldStop){
+            enabled=false;
+            shouldStop=false;
+            Stop();
+            return;
+        }
+
+        if(!enabled)
+            return;
+
+        std::lock_guard<std::mutex> lck (mtx);
+        samples.push_back(Point(FRAME,""));
+    }
+
+
+    /// Point saves the time it's constructed
+    Point::Point(Type type, const char *label=""): type(type), label(label){
+        auto timeElapsed = Sys::SteadyClock::now() - startTime;
+        time = std::chrono::duration_cast < std::chrono::microseconds > (timeElapsed).count();
+        tid = std::this_thread::get_id();
+    }
+
+
+    /// At the Beginning and End of every function
+
+    ProfilerGuard::ProfilerGuard( const char * label ): label( label ){
+        if(!enabled)
+            return;
+        localSamples.push_back( Point( START, label ) );
+    }
+
+    ProfilerGuard::~ProfilerGuard(){
+        if(!enabled)
+            return;
+        localSamples.push_back( Point( END, label ) );
+    }
+
+
+    void Stop(){
+        enabled=false;
+
+        auto file = FS::HomePath::OpenWrite("profile.log"); ///TODO add timestamp to file
+
+        std::sort(samples.begin(), samples.end());
```

Comment:
You'll need to lock this access too, to make the lock as tiny as possible, you can std:::move samples into a local variable.